daemon/containerd/soft_delete.go

```diff
@@ -0,0 +1,77 @@
+package containerd
+
+import (
+	"context"
+
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/errdefs"
+	"github.com/sirupsen/logrus"
+)
+
+// softImageDelete deletes the image, making sure that there are other images
+// that reference the content of the deleted image.
+// If no other image exists, a dangling one is created.
+func (i *ImageService) softImageDelete(img containerdimages.Image) error {
+	// Don't accept ctx as argument as we don't want this
+	// to be interrupted in the middle.
+	ctx := context.Background()
+	is := i.client.ImageService()
+
+	logger := logrus.WithFields(logrus.Fields{
+		"name":   img.Name,
+		"digest": img.Target.Digest.String(),
+	})
+
+	ref, err := reference.ParseNormalizedNamed(img.Name)
+	if err != nil {
+		logger.WithError(err).Debug("image name is not a valid reference")
+		return errdefs.InvalidParameter(err)
+	}
+
+	// If the image already exists, persist it as dangling image
+	// but only if no other image has the same target.
+	digest := img.Target.Digest.String()
+	imgs, err := is.List(ctx, "target.digest=="+digest)
+	if err != nil {
+		logger.WithField("digest", digest).WithError(err).Debug("failed to check if there are images targeting this digest")
+		return errdefs.System(err)
+	}
+
+	// Create dangling image if this is the last image pointing to this target.
+	if len(imgs) == 1 {
+		danglingImage := img
+
+		refWithoutTag := reference.TrimNamed(ref)
+		newRef, err := reference.WithDigest(refWithoutTag, img.Target.Digest)
```

Comment:
This will just discard the tag and still use the original repo.
So for example replacing `docker.io/library/busybox:latest` would change it to `docker.io/library/busybox@sha256:HASH`.
Wondering if we should drop the repo here.
WDYT?

Reply:
Looks like the current Docker drops the repo completely:
```bash
$ docker image ls | grep d3156fec8bcb
alpine-test                        latest                      d3156fec8bcb   8 weeks ago     7.46MB
$ docker tag alpine:3.17.1 alpine-test
$ docker image ls | grep d3156fec8bcb
<none>                             <none>                      d3156fec8bcb   8 weeks ago     7.46MB
$ docker inspect d3156fec8bcb
[
    {
        "Id": "sha256:d3156fec8bcbc7b491a4edc271a7734dcfa186fc73282d4e120eeaaf2ce95c43",
        "RepoTags": [],
        "RepoDigests": [],
...
```

Reply:
Changed, dangling image no longer references the old repository.